### PR TITLE
feat(textarea): fill and outline support

### PR DESCRIPTION
**Pull request by sean-perkins**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
  - Some docs updates need to be made in the `ionic-docs` repo, in a separate PR. See the [contributing guide](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#modifying-documentation) for details.
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Developers can not specify the shape or the fill of `ion-textarea` with the new syntax.

## What is the new behavior?

- Added template examples for outline/shape.
- Updated a11y tests to ensure new styles pass Axe criteria.
- Added screenshot tests for new styles.
- Added styles for rounded shape.
- Added style for `fill="solid"`.
- Added styles for `fill="outline"`.
- Added highlight styles for form controls to `ion-textarea`.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

👋 here is a Stackblitz of the legacy v6 templates with the same test layout as the new v7 examples: https://stackblitz.com/edit/qen13q?file=index.html. Expect small visual differences, callout things you are curious on.

### CODE REVIEW

File: core/src/components/textarea/textarea.md.outline.scss [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+ * by removing the top border from the notch fragment.
+ */
+:host(.has-focus.textarea-fill-outline.textarea-label-placement-floating) .textarea-outline-notch,
+:host(.has-value.textarea-fill-outline.textarea-label-placement-floating) .textarea-outline-notch,
+:host(.textarea-fill-outline.textarea-label-placement-stacked) .textarea-outline-notch {
+  border-top: none;
+}
+
+:host(.textarea-fill-outline) .native-textarea {
+  @include border-radius(0);
```
Comment (liamdebeasi) [inline core/src/components/textarea/textarea.md.outline.scss:218]: Should this be removed in the base textarea stylesheet?

File: core/src/components/textarea/test/fill/index.html [hunk 2]
```diff
@@ -0,0 +1,195 @@
+<!DOCTYPE html>
```
Comment (liamdebeasi) [inline core/src/components/textarea/test/fill/index.html:1]: Threading for discussion:

Thoughts on increasing the height of the textarea? The textarea height is so small that it looks like the text is off center. If you compare with https://material-components.github.io/material-components-web-catalog/#/component/text-field, the height is larger so the text does not look off center.

**Review by liamdebeasi — COMMENTED** [review]
Overall looks good. The height of the textarea makes it look

File: core/src/components/textarea/test/fill/index.html (hunk 2 above)
Comment (liamdebeasi) [reply core/src/components/textarea/test/fill/index.html:1]: Looking at https://material-components.github.io/material-components-web-catalog/#/component/text-field a  bit more, when you remove the height from the MD textarea, the dimensions and positioning look the same as their text field.

| default textarea | textarea with height removed |
| - | - |
| image | image |

I wonder if we'd have more luck if we tried this:

1. Revert textarea to have the same padding/margins as the new input syntax.
2. Make the textarea taller by default so it visually stands out when compared to the input.

**Comment by amandaejohnston** [comment]
The label appears slightly misaligned from the text (screenshot from the Label Placement test):
label misaligned

**Comment by sean-perkins** [comment]
CleanShot 2023-01-23 at 12 05 05

Looks to be off by a single pixel. I think I ran into that originally with the label placement PR. I'll review the recent changes and see what is contributing.

**Comment by liamdebeasi** [comment]
It's this commit: https://github.com/ionic-team/ionic-framework/pull/26610/commits/312c40c513a029cffce2b8740bdfa495b4bfdd52

I'm working on a fix now.

**Comment by liamdebeasi** [comment]
This should be ready for another review. A few things changes:

1. I set `--padding-bottom: 0px` instead of the non zero value it was at before. This was causing an issue where part of the text would look cutoff as the textarea was scrolled. This will result in a height change on the textarea in the screenshots.
2. Revamped how the textarea and label are aligned. In particular, I removed my old `max-height` logic and changed the `align-items` behavior from `stretch` to `start`: https://github.com/ionic-team/ionic-framework/pull/26610/commits/87814b161393f41d8b548d4b3bee67e6b23e3da7 The `start` change is necessary so that the label remains fixed to the top of the textarea as the textarea grows in size.

One of the challenges here (and main differences between textarea and input) is the textarea's text is aligned to the top of the editable area. With the input it is centered. This means we need to make up for this alignment diff elsewhere. As a result, there may be a 1-2px of alignment difference between the input and textarea due to this (and the fact that a textarea is typically taller than an input)

**Comment by amandaejohnston** [comment]
With the latest changes, now if you type more than a couple lines, you can see a random empty space above the textarea (note how the scrollbar on the right doesn't span the full height of the container):
empty space
This looks great with floating/stacked labels, but strange with the other placements. Looking at https://material-components.github.io/material-components-web-catalog/#/component/text-field, it seems only a floating label is officially spec'd out, so I would be okay with proceeding as-is if there's no way to cleanly fix this. I can imagine it's tricky to get the label aligned with the text while only adding top padding to the label itself, while accounting for cross-browser differences in how text is rendered.

**Comment by liamdebeasi** [comment]
Does this happen in with the old v6 textarea by any chance?

**Comment by amandaejohnston** [comment]
No, the text spans the container's full height:
full height

With only one or two lines of text, the label is vertically aligned:
aligned

**Comment by liamdebeasi** [comment]
I changed the implementation to apply the padding to the label and textarea directly instead of the wrapper. The spacing up top should be gone now

**Comment by sean-perkins** [comment]
Changes look good from my end based on the latest. @liamdebeasi do you feel similar, is this ready for merge?

**Comment by liamdebeasi** [comment]
Ship it!

**Pull request merged**